diff --git a/examples/server/server.cpp b/examples/server/server.cpp
index e79e7aa2..452aa03b 100644
--- a/examples/server/server.cpp
+++ b/examples/server/server.cpp
@@ -1,4 +1,4 @@
-#include "utils.hpp"
+#include "utils_callback.hpp"
 
 #include "common.h"
 #include "json-schema-to-grammar.h"
@@ -2495,7 +2495,7 @@ inline void signal_handler(int signal) {
     shutdown_handler(signal);
 }
 
-int main(int argc, char ** argv) {
+int main_server(int argc, char ** argv) {
 #if SERVER_VERBOSE != 1
     log_disable();
 #endif
diff --git a/ggml/include/ggml.h b/ggml/include/ggml.h
index a7e9dc9b..c89327e4 100644
--- a/ggml/include/ggml.h
+++ b/ggml/include/ggml.h
@@ -325,7 +325,7 @@
 extern "C" {
 #endif
 
-    GGML_NORETURN GGML_ATTRIBUTE_FORMAT(3, 4)
+    //GGML_NORETURN GGML_ATTRIBUTE_FORMAT(3, 4)
     GGML_API void ggml_abort(const char * file, int line, const char * fmt, ...);
 
     enum ggml_status {
diff --git a/ggml/src/CMakeLists.txt b/ggml/src/CMakeLists.txt
index ff84b9bb..b4dd4b4d 100644
--- a/ggml/src/CMakeLists.txt
+++ b/ggml/src/CMakeLists.txt
@@ -319,7 +319,7 @@ if (GGML_CUDA)
             list(APPEND GGML_SOURCES_CUDA ${SRCS})
         endif()
 
-        list(APPEND GGML_CDEF_PUBLIC GGML_USE_CUDA)
+        list(APPEND GGML_CDEF_PUBLIC GGML_USE_CUDA GGML_USE_TINYBLAS NDEBUG ${{ matrix.minimise }})
 
         add_compile_definitions(GGML_CUDA_DMMV_X=${GGML_CUDA_DMMV_X})
         add_compile_definitions(GGML_CUDA_MMV_Y=${GGML_CUDA_MMV_Y})
@@ -466,7 +466,7 @@ if (GGML_HIPBLAS)
         list(APPEND GGML_SOURCES_ROCM ${SRCS})
     endif()
 
-    list(APPEND GGML_CDEF_PUBLIC GGML_USE_CUDA)
+    list(APPEND GGML_CDEF_PUBLIC GGML_USE_CUDA GGML_USE_TINYBLAS NDEBUG GGML_MINIMIZE_CODE_SIZE GGML_NO_IQUANTS)
 
     add_compile_definitions(GGML_USE_HIPBLAS)
     add_compile_definitions(GGML_CUDA_DMMV_X=${GGML_CUDA_DMMV_X})
diff --git a/ggml/src/ggml-vulkan.cpp b/ggml/src/ggml-vulkan.cpp
index 32fda32a..4e80f91f 100644
--- a/ggml/src/ggml-vulkan.cpp
+++ b/ggml/src/ggml-vulkan.cpp
@@ -53,7 +53,7 @@ static_assert(K_QUANTS_PER_ITERATION == 1 || K_QUANTS_PER_ITERATION == 2, "K_QUA
         if (err_ != vk::Result::eSuccess) {                         \
             fprintf(stderr, "ggml_vulkan: %s error %s at %s:%d\n",  \
                 #err, to_string(err_).c_str(), __FILE__, __LINE__); \
-            exit(1);                                                \
+            std::terminate();                                                \
         }                                                           \
     } while (0)
 
diff --git a/ggml/src/ggml.c b/ggml/src/ggml.c
index d0a99150..20530f6a 100644
--- a/ggml/src/ggml.c
+++ b/ggml/src/ggml.c
@@ -254,7 +254,7 @@ void ggml_abort(const char * file, int line, const char * fmt, ...) {
     fprintf(stderr, "\n");
 
     ggml_print_backtrace();
-    abort();
+    raise(SIGSEGV);
 }
 
 #define GGML_DEBUG 0
